### ISSUE TITLE
Add possibility to plot holes in plolygons (support for `grid.path()`)

### ISSUE THREAD

**Issue by courtiol** [deepayan/lattice#36]
As discussed elsewhere (https://github.com/oscarperpinan/rastervis/issues/101) it would be great to add native support in lattice for drawing polygons with holes.

FYI, here is how they did it in *sp*:
https://github.com/edzer/sp/blob/fe2993112e1d7d6f0c56b8595041f0a5ee5ef895/R/spplot.R#L1-L47

**Comment by deepayan** [comment]
Sure, let me try...

**Comment by deepayan** [comment]
I have added a `rule` argument to `lpolygon.default()`; check if this works for you. You will need to specify `rule="evenodd"` or `rule="winding"`. For the example in <https://github.com/oscarperpinan/rastervis/issues/101>, the following works:

```r
rasterVis::levelplot(worldelev_terra, margin = FALSE) +
  latticeExtra::layer(lpolygon.SpatVector(oceans_terra, border = "yellow", 
                                          col = "blue", rule = "winding"))
```

This is slow for the on-screen quartz device, but fast for pdf.

**Comment by courtiol** [comment]
Wonderful, it seems to work fine on my side.

One open question for me is what the default value for `rule` should be?
I understand that from a theoretical backward compatibility perspective, `"none"` is the right choice.
Yet, I don't know if in practice anyone would combine polygons within the same shape if not to draw holes (in such a case  `"evenodd"` would be more convenient)...
I guess we will never be sure and should perhaps stick to `"none"`.

In any case, perhaps, to better illustrate the `rule` argument, it would be great to add the following examples adapted from Murrell's paper (https://journal.r-project.org/archive/2012-2/RJournal_2012-2_Murrell2.pdf):

``` r
x <- c(.1, .5, .9, NA,
       .4, .5, .6, NA,
       .4, .6, .5)

y <- c(.1, .8, .1, NA,
       .5, .4, .5, NA,
       .3, .3, .2)

d <- data.frame(x = x, y = y)

lattice::xyplot(y ~ x, data = d, panel = lattice::panel.polygon, rule = "none", col = "grey")
```

![](https://i.imgur.com/4YWgLuE.png)

``` r
lattice::xyplot(y ~ x, data = d, panel = lattice::panel.polygon, rule = "evenodd", col = "grey")
```

![](https://i.imgur.com/s10x0ok.png)

``` r
lattice::xyplot(y ~ x, data = d, panel = lattice::panel.polygon, rule = "winding", col = "grey")
```

![](https://i.imgur.com/sZVxcCg.png)

<sup>Created on 2023-10-11 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Thanks again!

**Comment by courtiol** [comment]
Shall we define `lpolypath()` as a wrapper around  `lpolygon.default()` with `rule = "evenodd"` as default?
I am happy to PR that while adding the example above if that helps.

**Comment by deepayan** [comment]
No need for a PR, it's going to be trivial... The default of `rule = "evenodd"` would then be different from the default in `polypath()`, but I guess that's OK.

**Comment by courtiol** [comment]
My 2 cts...

I have little experience about the effect of the rules for practical applications.

From the examples I found, `"evenodd"` seems to reveal more holes, while `"winding"` (called `"nonzero"` below) seems to keep polygons pilling on top of each others (see e.g. section **More background** below).

In "sp", they used `"winding"` as default:
https://github.com/edzer/sp/blob/fe2993112e1d7d6f0c56b8595041f0a5ee5ef895/R/AAA.R#L9

And as you said, in `polypath()` they did the same.
So I think it would be best for you to stick to `"winding"` as default for `lpolypath()`.

If that is not best for "rasterVis" applications (but I think it makes no difference in most cases), we can always change the default there, but "lattice" being a package for widespread use, I would be consistent with base R choices.

### More background
From https://www.w3.org/TR/SVG/painting.html#FillRuleProperty:

![image](https://github.com/deepayan/lattice/assets/7499666/6e02ec36-3779-42dd-8f04-2ddc8d963c9e)

**Comment by deepayan** [comment]
Yes, I think it's best to be consistent. I have now added `lpolypath()` etc.

**Comment by courtiol** [comment]
Ok, I think you can now close this issue.
Do you already know your time line before the next CRAN release of lattice?
Many thanks again!

**Comment by deepayan** [comment]
I'm hoping to do a release 'soon'. Let's keep this open till then.
